### PR TITLE
[JAMES-2086] FIX: Handle OverQuotaException instead of more generic MailboxException

### PR DESCRIPTION
**Pull request by FilipRazek**

### CODE REVIEW

File: mailbox/api/src/main/java/org/apache/james/mailbox/quota/QuotaManager.java [hunk 1]
```diff
@@ -57,7 +57,7 @@ public Quota<QuotaSizeLimit, QuotaSizeUsage> getStorageQuota() {
      *
      * @param quotaRoot Quota root argument from RFC 2087 ( correspond to the user owning this mailbox )
      */
-    Quota<QuotaCountLimit, QuotaCountUsage> getMessageQuota(QuotaRoot quotaRoot) throws MailboxException;
+    Quota<QuotaCountLimit, QuotaCountUsage> getMessageQuota(QuotaRoot quotaRoot) throws OverQuotaException;
```
Comment (chibenwa) [inline mailbox/api/src/main/java/org/apache/james/mailbox/quota/QuotaManager.java:60]: You do not get a OverQuotaException by simply checking your quota, this do not make sense

File: protocols/imap/src/main/java/org/apache/james/imap/processor/AbstractMessageRangeProcessor.java [hunk 2]
```diff
@@ -82,7 +83,12 @@ protected Mono<Void> processRequestReactive(R request, ImapSession session, Resp
             .onErrorResume(MessageRangeException.class, e -> {
                 taggedBad(request, responder, HumanReadableText.INVALID_MESSAGESET);
                 return ReactorUtils.logAsMono(() -> LOGGER.debug("{} failed from mailbox {} to {} for invalid sequence-set {}",
-                    getOperationName(), session.getSelected().getMailboxId(), targetMailbox, request.getIdSet(), e));
+                        getOperationName(), session.getSelected().getMailboxId(), targetMailbox, request.getIdSet(), e));
+            })
+            .onErrorResume(OverQuotaException.class, e -> {
+                taggedBad(request, responder, HumanReadableText.FAILURE_OVERQUOTA);
+                return ReactorUtils.logAsMono(() -> LOGGER.debug("{} failed from mailbox {} to {} for invalid sequence-set {}",
```
Comment (chibenwa) [inline protocols/imap/src/main/java/org/apache/james/imap/processor/AbstractMessageRangeProcessor.java:90]: info log.

Comment (chibenwa) [reply protocols/imap/src/main/java/org/apache/james/imap/processor/AbstractMessageRangeProcessor.java:90]: And I bet there is nothing wrong with the sequence-set

File: protocols/imap/src/main/java/org/apache/james/imap/processor/AbstractMailboxProcessor.java [hunk 3]
```diff
@@ -100,6 +101,10 @@ protected final Mono<Void> doProcess(R acceptableMessage, Responder responder, I
                             no(acceptableMessage, responder, HumanReadableText.DENIED_SHARED_MAILBOX);
                             return Mono.empty();
                         })
+                        .onErrorResume(OverQuotaException.class, e -> {
+                            no(acceptableMessage, responder, HumanReadableText.FAILURE_OVERQUOTA);
+                            return Mono.empty();
```
Comment (chibenwa) [inline protocols/imap/src/main/java/org/apache/james/imap/processor/AbstractMailboxProcessor.java:106]: Maybe an INFO log could be a good idea?

Comment (chibenwa) [reply protocols/imap/src/main/java/org/apache/james/imap/processor/AbstractMailboxProcessor.java:106]: BTW we can add an overQuota response code, cf https://datatracker.ietf.org/doc/html/rfc5530

Quote from `AppendProcessor`: `no(request, responder, HumanReadableText.FAILURE_OVERQUOTA, StatusResponse.ResponseCode.overQuota());`

File: server/container/mailbox-jmx/src/main/java/org/apache/james/adapter/mailbox/MailboxManagerManagement.java [hunk 4]
```diff
@@ -200,6 +201,8 @@ public long size() {
                         return file.length();
                     }
                 }), session);
+        } catch (OverQuotaException e) {
+            LOGGER.error("Unable to create mailbox due to quota error", e);
```
Comment (chibenwa) [inline server/container/mailbox-jmx/src/main/java/org/apache/james/adapter/mailbox/MailboxManagerManagement.java:205]: Log messages is not accurate.

I suggest: Unable to import due to quota error

File: server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/http/JMAPApiRoutes.java [hunk 5]
```diff
@@ -102,6 +103,7 @@ private Mono<Void> post(HttpServerRequest request, HttpServerResponse response)
                 .contextWrite(jmapAuthContext(session)))
             .onErrorResume(BadRequestException.class, e -> handleBadRequest(response, LOGGER, e))
             .onErrorResume(UnauthorizedException.class, e -> handleAuthenticationFailure(response, LOGGER, e))
+            .onErrorResume(OverQuotaException.class, e -> handleBadRequest(response, LOGGER, e))
```
Comment (chibenwa) [inline server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/http/JMAPApiRoutes.java:106]: I'm pretty sure the fact that it fails have nothing related with the syntax of the client...

We should handle this within the dedicated class: SetMessagesCreationProcessor and SetMessagesUpdateProcessor, catch the overQuotaException there and handle it.

**Review by chibenwa — COMMENTED** [review]
Hello,

Thanks for proposing a fix, I ave 2 broad remark:

 - error management is hard, we need to pay attention of handling it in the right place, with the right approach, with the right log. Failure to do this properly is a guaranty for debugging headaches. In this regard we can clearly improve this work.
 - This pull request do not test anything and as such we should rework it.
 
I must thank you because reviewing in detail the code here made me realise that the error handling code for IMAP COPY could greatly be improved:

```
A008 NO COPY processing failed.
=>
A008 NO [OVERQUOTA] COPY failed. Over quota.
```

Regards,

Benoit

File: protocols/imap/src/main/java/org/apache/james/imap/processor/AbstractMessageRangeProcessor.java [hunk 6]
```diff
@@ -84,6 +85,11 @@ protected Mono<Void> processRequestReactive(R request, ImapSession session, Resp
                 return ReactorUtils.logAsMono(() -> LOGGER.debug("{} failed from mailbox {} to {} for invalid sequence-set {}",
                     getOperationName(), session.getSelected().getMailboxId(), targetMailbox, request.getIdSet(), e));
             })
+            .onErrorResume(OverQuotaException.class, e -> {
+                taggedBad(request, responder, HumanReadableText.FAILURE_OVERQUOTA);
```
Comment (chibenwa) [inline protocols/imap/src/main/java/org/apache/james/imap/processor/AbstractMessageRangeProcessor.java:89]: Should be a no, should have OVERQUOTA response code.

File: protocols/imap/src/main/java/org/apache/james/imap/processor/AbstractMailboxProcessor.java [hunk 7]
```diff
@@ -100,6 +101,11 @@ protected final Mono<Void> doProcess(R acceptableMessage, Responder responder, I
                             no(acceptableMessage, responder, HumanReadableText.DENIED_SHARED_MAILBOX);
                             return Mono.empty();
                         })
+                        .onErrorResume(OverQuotaException.class, e -> {
+                            no(acceptableMessage, responder, HumanReadableText.FAILURE_OVERQUOTA);
```
Comment (chibenwa) [inline protocols/imap/src/main/java/org/apache/james/imap/processor/AbstractMailboxProcessor.java:105]: Idem let's use OVERQUOTA response code?

File: server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/http/JMAPApiRoutes.java (hunk 5 above)
Comment (chibenwa) [reply server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/http/JMAPApiRoutes.java]: More line 264 as the update on the mailboxIds can be used to copy messages and thus might trigger overquota exceptions ;-)

File: protocols/imap/src/main/java/org/apache/james/imap/processor/AbstractMessageRangeProcessor.java [hunk 8]
```diff
@@ -86,7 +86,7 @@ protected Mono<Void> processRequestReactive(R request, ImapSession session, Resp
                     getOperationName(), session.getSelected().getMailboxId(), targetMailbox, request.getIdSet(), e));
             })
             .onErrorResume(OverQuotaException.class, e -> {
-                taggedBad(request, responder, HumanReadableText.FAILURE_OVERQUOTA);
+                no(request, responder, HumanReadableText.FAILURE_OVERQUOTA);
```
Comment (chibenwa) [inline protocols/imap/src/main/java/org/apache/james/imap/processor/AbstractMessageRangeProcessor.java:89]:
```suggestion
                no(request, responder, HumanReadableText.FAILURE_OVERQUOTA, StatusResponse.ResponseCode.overQuota());
```

Here and in other places.

File: protocols/imap/src/main/java/org/apache/james/imap/processor/AbstractMessageRangeProcessor.java (hunk 6 above)
Comment (FilipRazek) [reply protocols/imap/src/main/java/org/apache/james/imap/processor/AbstractMessageRangeProcessor.java]: Updated

File: protocols/imap/src/main/java/org/apache/james/imap/processor/AbstractMailboxProcessor.java (hunk 7 above)
Comment (FilipRazek) [reply protocols/imap/src/main/java/org/apache/james/imap/processor/AbstractMailboxProcessor.java]: Done :+1:

File: protocols/imap/src/main/java/org/apache/james/imap/processor/AbstractMailboxProcessor.java [hunk 9]
```diff
@@ -102,7 +102,7 @@ protected final Mono<Void> doProcess(R acceptableMessage, Responder responder, I
                             return Mono.empty();
                         })
                         .onErrorResume(OverQuotaException.class, e -> {
-                            no(acceptableMessage, responder, HumanReadableText.FAILURE_OVERQUOTA);
+                            no(acceptableMessage, responder, HumanReadableText.FAILURE_OVERQUOTA, StatusResponse.ResponseCode.overQuota());
                             LOGGER.info("Processing failed due to quota restriction");
```
Comment (chibenwa) [inline protocols/imap/src/main/java/org/apache/james/imap/processor/AbstractMailboxProcessor.java:106]: Ideally we should use `ReactorUtils.logAsMono(() -> LOGGER.info(...))` syntax in order not to loose the MDC context

**Review by chibenwa — APPROVED** [review]

**Review by Arsnael — APPROVED** [review]
Thank you for your contribution @FilipRazek  :)

**Pull request merged**